File: cdk/examples/data_pipeline/typescript/lib/data-pipeline-roles.ts
import { Effect, ManagedPolicy, Policy, PolicyStatement, Role } from "aws-cdk-lib/aws-iam";

export function addStepFunctionRolePolicies(account: String, region: String, stepFunctionExecutionRole: Role) {
    stepFunctionExecutionRole.addToPrincipalPolicy(new PolicyStatement({
        actions:["ecs:RunTask"],
        effect: Effect.ALLOW,
        resources: [`arn:aws:ecs:${region}:${account}:task-definition/*`]
    }))
    stepFunctionExecutionRole.addToPrincipalPolicy(new PolicyStatement({
        actions: ["logs:CreateLogDelivery",
        "logs:GetLogDelivery",
        "logs:UpdateLogDelivery",
        "logs:DeleteLogDelivery",
        "logs:ListLogDeliveries",
        "logs:PutResourcePolicy",
        "logs:DescribeResourcePolicies",
        "logs:DescribeLogGroups"],
        effect: Effect.ALLOW,
        resources: ["*"]
    }))
    stepFunctionExecutionRole.addToPrincipalPolicy(new PolicyStatement({
        actions: ["xray:PutTraceSegments",
        "xray:PutTelemetryRecords",
        "xray:GetSamplingRules",
        "xray:GetSamplingTargets"],
        effect: Effect.ALLOW,
        resources: ["*"]
    }))
}

export function addEcsTaskExecutionRolePolicies(account: String, region: String, ecsTaskExecutionRole: Role) {
    ecsTaskExecutionRole.addToPrincipalPolicy(new PolicyStatement({
        actions: ["ecr:GetAuthorizationToken",
        "ecr:BatchCheckLayerAvailability",
        "ecr:GetDownloadUrlForLayer",
        "ecr:BatchGetImage"],
        effect: Effect.ALLOW,
        resources: [`arn:aws:ecr:${region}:${account}:*`]
    }))
    ecsTaskExecutionRole.addToPrincipalPolicy(new PolicyStatement({
        actions: ["logs:CreateLogStream",
        "logs:PutLogEvents"],
        effect: Effect.ALLOW,
        resources: ["*"]
    }))
    ecsTaskExecutionRole.addToPrincipalPolicy(new PolicyStatement({
        actions: ["ec2:AuthorizeSecurityGroupIngress",
        "ec2:Describe*"],
        effect: Effect.ALLOW,
        resources: [`arn:aws:ec2:${region}:${account}:*`]
    }))
    ecsTaskExecutionRole.addToPrincipalPolicy(new PolicyStatement({
        actions: ["elasticloadbalancing:DeregisterInstancesFromLoadBalancer",
        "elasticloadbalancing:DeregisterTargets",
        "elasticloadbalancing:Describe*",
        "elasticloadbalancing:RegisterInstancesWithLoadBalancer",
        "elasticloadbalancing:RegisterTargets"],
        effect: Effect.ALLOW,
        resources: [`arn:aws:elasticloadbalancing:${region}:${account}:*`]
    }))
}

export function addEcsTaskRolePolicies(account: String, region: String, ecsTaskRole: Role) {
    ecsTaskRole.addToPrincipalPolicy(new PolicyStatement({
        actions: ["states:*"],
        effect: Effect.ALLOW,
        resources: [`arn:aws:states:${region}:${account}:*`]
    }))
    ecsTaskRole.addToPrincipalPolicy(new PolicyStatement({
        actions: ["ssm:DescribeParameters",
        "ssm:GetParameters"],
        effect: Effect.ALLOW,
        resources:[`arn:aws:ssm:${region}:${account}:*`]
    }))
    ecsTaskRole.addToPrincipalPolicy(new PolicyStatement({
        actions: ["kms:Decrypt"],
        effect: Effect.ALLOW,
        resources: [`arn:aws:kms:${region}:${account}:*`]
    }))
    ecsTaskRole.addToPrincipalPolicy(new PolicyStatement({
        actions: ["elasticfilesystem:*"],
        effect: Effect.ALLOW,
        resources: [`arn:aws:elasticfilesystem:${region}:${account}:*`]
    }))
    ecsTaskRole.addToPrincipalPolicy(new PolicyStatement({
        actions: ["s3:*",
        "s3-object-lambda:*"],
        effect: Effect.ALLOW,
        resources: [`arn:aws:s3:::*`]
    }))
}

export function addLambdaExecutionRolePolicies(account: String, region: String, lambdaExecutionRole: Role) {
    lambdaExecutionRole.addToPrincipalPolicy(new PolicyStatement({
        actions: ["kms:Decrypt"],
        effect: Effect.ALLOW,
        resources: ["*"]
    }))
    lambdaExecutionRole.addToPrincipalPolicy(new PolicyStatement({
        actions: ["s3:*",
        "s3-object-lambda:*"],
        effect: Effect.ALLOW,
        resources: [`arn:aws:s3:::*`]
    }))
}

File: cdk/examples/data_pipeline/python/lib/data_pipeline_roles.py
import os
import aws_cdk.aws_iam as iam
from lib.data_pipeline_stack_props import DataPipelineStackProps

def add_step_function_role_policies(stepfunctionExecutionRole:iam.Role, data_pipeline_stack_props: DataPipelineStackProps):
    stepfunctionExecutionRole.add_to_principal_policy(iam.PolicyStatement(
        actions= ['ecs:RunTask'],
        effect= iam.Effect.ALLOW,
        resources= ['arn:aws:ecs:'+data_pipeline_stack_props.aws_region+':'+data_pipeline_stack_props.account_number+':task-definition/*']
    ))
    stepfunctionExecutionRole.add_to_principal_policy(iam.PolicyStatement(
        actions= ["logs:CreateLogDelivery",
        "logs:GetLogDelivery",
        "logs:UpdateLogDelivery",
        "logs:DeleteLogDelivery",
        "logs:ListLogDeliveries",
        "logs:PutResourcePolicy",
        "logs:DescribeResourcePolicies",
        "logs:DescribeLogGroups"],
        effect= iam.Effect.ALLOW,
        resources= ["*"]
    ))
    stepfunctionExecutionRole.add_to_principal_policy(iam.PolicyStatement(
        actions= ["xray:PutTraceSegments",
        "xray:PutTelemetryRecords",
        "xray:GetSamplingRules",
        "xray:GetSamplingTargets"],
        effect= iam.Effect.ALLOW,
        resources= ["*"]
    ))

    return stepfunctionExecutionRole

def add_ecs_task_execution_role_policies(ecsTaskExecutionRole:iam.Role, data_pipeline_stack_props: DataPipelineStackProps):
    ecsTaskExecutionRole.add_to_principal_policy(iam.PolicyStatement(
        actions= ["ecr:GetAuthorizationToken",
        "ecr:BatchCheckLayerAvailability",
        "ecr:GetDownloadUrlForLayer",
        "ecr:BatchGetImage"],
        effect= iam.Effect.ALLOW,
        resources= ['arn:aws:ecr:'+data_pipeline_stack_props.aws_region+':'+data_pipeline_stack_props.account_number+':*']
    ))
    ecsTaskExecutionRole.add_to_principal_policy(iam.PolicyStatement(
        actions= ["logs:CreateLogStream",
        "logs:PutLogEvents"],
        effect= iam.Effect.ALLOW,
        resources= ["*"]
    ))
    ecsTaskExecutionRole.add_to_principal_policy(iam.PolicyStatement(
        actions= ["ec2:AuthorizeSecurityGroupIngress",
        "ec2:Describe*"],
        effect= iam.Effect.ALLOW,
        resources= ['arn:aws:ec2:'+data_pipeline_stack_props.aws_region+':'+data_pipeline_stack_props.account_number+':*']
    ))
    ecsTaskExecutionRole.add_to_principal_policy(iam.PolicyStatement(
        actions= ["elasticloadbalancing:DeregisterInstancesFromLoadBalancer",
        "elasticloadbalancing:DeregisterTargets",
        "elasticloadbalancing:Describe*",
        "elasticloadbalancing:RegisterInstancesWithLoadBalancer",
        "elasticloadbalancing:RegisterTargets"],
        effect= iam.Effect.ALLOW,
        resources= ['arn:aws:elasticloadbalancing:'+data_pipeline_stack_props.aws_region+':'+data_pipeline_stack_props.account_number+':*']
    ))

    return ecsTaskExecutionRole

def add_ecs_task_role_policies(ecsTaskRole: iam.Role, data_pipeline_stack_props: DataPipelineStackProps):
    ecsTaskRole.add_to_principal_policy(iam.PolicyStatement(
        actions= ["states:*"],
        effect= iam.Effect.ALLOW,
        resources= ['arn:aws:states:'+data_pipeline_stack_props.aws_region+':'+data_pipeline_stack_props.account_number+':*']
    ))
    ecsTaskRole.add_to_principal_policy(iam.PolicyStatement(
        actions= ["ssm:DescribeParameters",
        "ssm:GetParameters"],
        effect= iam.Effect.ALLOW,
        resources=['arn:aws:ssm:'+data_pipeline_stack_props.aws_region+':'+data_pipeline_stack_props.account_number+':*']
    ))
    ecsTaskRole.add_to_principal_policy(iam.PolicyStatement(
        actions= ["kms:Decrypt"],
        effect= iam.Effect.ALLOW,
        resources= ['arn:aws:kms:'+data_pipeline_stack_props.aws_region+':'+data_pipeline_stack_props.account_number+':*']
    ))
    ecsTaskRole.add_to_principal_policy(iam.PolicyStatement(
        actions= ["elasticfilesystem:*"],
        effect= iam.Effect.ALLOW,
        resources= ['arn:aws:elasticfilesystem:'+data_pipeline_stack_props.aws_region+':'+data_pipeline_stack_props.account_number+':*']
    ))
    ecsTaskRole.add_to_principal_policy(iam.PolicyStatement(
        actions= ["s3:*",
        "s3-object-lambda:*"],
        effect= iam.Effect.ALLOW,
        resources= ['arn:aws:s3:::*']
    ))

    return ecsTaskRole

def add_lambda_execution_role_policies(lambdaExecutionRole: iam.Role):
    lambdaExecutionRole.add_to_principal_policy(iam.PolicyStatement(
        actions= ["kms:Decrypt"],
        effect= iam.Effect.ALLOW,
        resources= ["*"]
    ))
    lambdaExecutionRole.add_to_principal_policy(iam.PolicyStatement(
        actions= ["s3:*",
        "s3-object-lambda:*"],
        effect= iam.Effect.ALLOW,
        resources= ['arn:aws:s3:::*']
    ))

    return lambdaExecutionRole
